File: src/main/java/io/github/Earth1283/aptMc/commands/sub/InstallCommand.java
package io.github.Earth1283.aptMc.commands.sub;

import io.github.Earth1283.aptMc.AptMc;
import io.github.Earth1283.aptMc.api.HangarAPI;
import io.github.Earth1283.aptMc.api.ModrinthAPI;
import io.github.Earth1283.aptMc.commands.SubCommand;
import io.github.Earth1283.aptMc.listeners.ConfirmationListener;
import io.github.Earth1283.aptMc.managers.PackageManager;
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder;
import org.bukkit.command.CommandSender;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import java.io.File;
import java.util.*;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;

public class InstallCommand extends SubCommand {
    private final ConfirmationListener confirmationListener;

    public InstallCommand(AptMc plugin, PackageManager packageManager, ConfirmationListener confirmationListener) {
        super(plugin, packageManager);
        this.confirmationListener = confirmationListener;
    }

    @Override
    public void execute(CommandSender sender, List<String> args, boolean dryRun) {
        if (args.isEmpty()) {
            sender.sendMessage(plugin.getMessage("errors.no-packages-specified"));
            return;
        }

        // Remote install check
        if (args.size() == 1 && args.get(0).matches("^(http|https|ftp)://.*\\.yml$")) {
            String url = args.get(0);
            if (dryRun) {
                sender.sendMessage(plugin.getMessage("status.dry-run", Placeholder.unparsed("arg", "Import manifest from " + url)));
                return;
            }
            try {
                sendStatus(sender, plugin.getMessage("status.downloading", Placeholder.unparsed("arg", "manifest")));
                File tempDir = new File(plugin.getDataFolder(), "temp");
                if (!tempDir.exists()) tempDir.mkdirs();
                File tempFile = new File(tempDir, "remote_import_" + System.currentTimeMillis() + ".yml");
                
                // Use HttpClient for robust downloading with User-Agent
                java.net.http.HttpClient client = java.net.http.HttpClient.newHttpClient();
                java.net.http.HttpRequest request = java.net.http.HttpRequest.newBuilder()
                        .uri(java.net.URI.create(url))
                        .header("User-Agent", "apt-mc/1.2 (parody-cli)")
                        .GET()
                        .build();

                java.net.http.HttpResponse<java.nio.file.Path> response = client.send(request, java.net.http.HttpResponse.BodyHandlers.ofFile(tempFile.toPath()));
                
                if (response.statusCode() != 200) {
                    sender.sendMessage(plugin.getMessage("errors.fetch-manifest-failed", Placeholder.unparsed("error", "HTTP " + response.statusCode())));
                    return;
                }

                if (!tempFile.exists() || tempFile.length() == 0) {
                    sender.sendMessage(plugin.getMessage("errors.fetch-manifest-failed", Placeholder.unparsed("error", "File empty or missing")));
                    return;
                }
                
                FileConfiguration yaml = YamlConfiguration.loadConfiguration(tempFile);
                String title = yaml.getString("project-details.title", "Unknown");
                String author = yaml.getString("project-details.author", "Unknown");
                
                sender.sendMessage(plugin.getMessage("status.import-header", Placeholder.unparsed("title", title), Placeholder.unparsed("author", author)));
                sender.sendMessage(plugin.getMessage("status.confirm-prompt"));
                
                if (confirmationListener != null) {
                    confirmationListener.addPending(sender, tempFile);
                } else {
                    sender.sendMessage(plugin.getMessage("errors.listener-missing"));
                }
                
            } catch (Exception e) {
                sender.sendMessage(plugin.getMessage("errors.fetch-manifest-failed", Placeholder.unparsed("error", e.getMessage())));
            }
            return;
        }

        packageManager.ensureDir();
        sendStatus(sender, plugin.getMessage("status.update-reading"));
        sendStatus(sender, plugin.getMessage("status.update-building"));

        Set<String> installedIds = getInstalledProjectIds();
        Map<String, JsonObject> toInstallMap = new HashMap<>();
        List<String> priority = plugin.getConfig().getStringList("source-priority");
        if (priority.isEmpty()) priority = List.of("modrinth", "hangar", "github");

        for (String pkgSlug : args) {
            boolean found = false;
            try {
                // Check if it's a GitHub slug (owner/repo)
                if (pkgSlug.contains("/") && !pkgSlug.startsWith("http")) {
                    sendStatus(sender, plugin.getMessage("status.checking", Placeholder.unparsed("arg", pkgSlug + " on GitHub")));
                    JsonObject release = io.github.Earth1283.aptMc.api.GitHubAPI.getLatestRelease(pkgSlug);
                    if (release != null) {
                        JsonObject asset = io.github.Earth1283.aptMc.api.GitHubAPI.getPrimaryJarAsset(release);
                        if (asset != null) {
                            JsonObject githubInfo = new JsonObject();
                            githubInfo.addProperty("source", "github");
                            githubInfo.addProperty("slug", pkgSlug);
                            githubInfo.addProperty("download_url", asset.get("browser_download_url").getAsString());
                            githubInfo.addProperty("filename", asset.get("name").getAsString());
                            githubInfo.addProperty("size", asset.get("size").getAsLong());
                            toInstallMap.put("github:" + pkgSlug, githubInfo);
                            found = true;
                        }
                    }
                }

                if (!found) {
                    for (String source : priority) {
                        if (source.equalsIgnoreCase("modrinth")) {
                            sendStatus(sender, plugin.getMessage("status.checking", Placeholder.unparsed("arg", pkgSlug + " on Modrinth")));
                            JsonObject project = ModrinthAPI.getProject(pkgSlug);
                            if (project != null) {
                                resolveDependenciesRecursive(sender, project, toInstallMap, installedIds);
                                found = true;
                                break;
                            }
                        } else if (source.equalsIgnoreCase("hangar") && plugin.getConfig().getBoolean("enable-hangar", true)) {
                            sendStatus(sender, plugin.getMessage("status.checking", Placeholder.unparsed("arg", pkgSlug + " on Hangar")));
                            JsonObject hangarProject = HangarAPI.getProject(pkgSlug);
                            if (hangarProject != null) {
                                JsonObject versionInfo = HangarAPI.getLatestVersion(pkgSlug, "PAPER");
                                if (versionInfo != null) {
                                    versionInfo.addProperty("source", "hangar");
                                    versionInfo.addProperty("hangar_slug", hangarProject.get("name").getAsString());
                                    toInstallMap.put("hangar:" + hangarProject.get("name").getAsString(), versionInfo);
                                    found = true;
                                    break;
                                }
                            }
                        }
                    }
                }
                
                if (!found) {
                    sender.sendMessage(plugin.getMessage("errors.package-not-found", Placeholder.unparsed("arg", pkgSlug)));
                }
            } catch (Exception e) {
                sender.sendMessage(plugin.getMessage("errors.error-checking-pkg", Placeholder.unparsed("arg", pkgSlug), Placeholder.unparsed("error", e.getMessage())));
            }
        }

        List<JsonObject> toInstall = new ArrayList<>(toInstallMap.values());
        if (toInstall.isEmpty()) {
            sender.sendMessage(plugin.getMessage("status.up-to-date")); // Generic "nothing to do"
            return;
        }

        if (dryRun) {
            sender.sendMessage(plugin.getMessage("status.dry-run-header"));
            for (JsonObject pkg : toInstall) {
                String name = getPkgName(pkg);
                sender.sendMessage(plugin.getMessage("status.install-item", Placeholder.unparsed("arg", name)));
            }
            sender.sendMessage(plugin.getMessage("status.dry-run-complete"));
            return;
        }

        if (plugin.getConfig().getBoolean("apt-song-references")) {
            sender.sendMessage(plugin.getMessage("status.install-song-ref")); 
        }

        sender.sendMessage(plugin.getMessage("status.install-header"));
        for (JsonObject pkg : toInstall) {
            String name = getPkgName(pkg);
            sender.sendMessage(plugin.getMessage("status.install-item", Placeholder.unparsed("arg", name)));
        }
        sender.sendMessage(plugin.getMessage("status.install-summary", Placeholder.unparsed("arg", String.valueOf(toInstall.size()))));

        List<CompletableFuture<Void>> installFutures = toInstall.stream()
                .map(pkg -> CompletableFuture.runAsync(() -> {
                    String source = pkg.has("source") ? pkg.get("source").getAsString() : "modrinth";
                    
                    if (source.equals("github")) {
                        try {
                            String url = pkg.get("download_url").getAsString();
                            String filename = pkg.get("filename").getAsString();
                            long size = pkg.get("size").getAsLong();
                            
                            sendStatus(sender, plugin.getMessage("status.downloading", Placeholder.unparsed("arg", filename)));
                            packageManager.downloadFile(url, packageManager.getPluginsDir(), filename, size, createProgressCallback(sender, filename));
                            sendStatus(sender, plugin.getMessage("status.downloaded", Placeholder.unparsed("arg", filename)));
                        } catch (Exception e) {
                             sender.sendMessage(plugin.getMessage("errors.install-failed", Placeholder.unparsed("arg", pkg.get("slug").getAsString()), Placeholder.unparsed("error", e.getMessage())));
                        }
                        return;
                    }

                    if (source.equals("hangar")) {
                        try {
                            String url = HangarAPI.getDownloadUrl(pkg, "PAPER");
                            String filename = HangarAPI.getFileName(pkg, "PAPER");
                            
                            if (url == null || filename == null) {
                                 sender.sendMessage(plugin.getMessage("errors.hangar-resolve-failed", Placeholder.unparsed("arg", pkg.get("hangar_slug").getAsString())));
                                 return;
                            }
                            
                            sendStatus(sender, plugin.getMessage("status.downloading", Placeholder.unparsed("arg", filename)));
                            packageManager.downloadFile(url, packageManager.getPluginsDir(), filename, -1, createProgressCallback(sender, filename));
                            sendStatus(sender, plugin.getMessage("status.downloaded", Placeholder.unparsed("arg", filename)));
                        } catch (Exception e) {
                             sender.sendMessage(plugin.getMessage("errors.install-failed", Placeholder.unparsed("arg", "Hangar Package"), Placeholder.unparsed("error", e.getMessage())));
                        }
                        return;
                    }

                    String slug = pkg.get("slug").getAsString();
                    try {
                        JsonArray versions = ModrinthAPI.getVersions(pkg.get("id").getAsString(), List.of("spigot", "paper", "purpur", "bukkit"));
                        if (versions.size() == 0) {
                            sender.sendMessage(plugin.getMessage("errors.install-failed", Placeholder.unparsed("arg", slug), Placeholder.unparsed("error", "No compatible versions")));
                            return;
                        }
                        
                        JsonObject latest = versions.get(0).getAsJsonObject();
                        JsonArray files = latest.getAsJsonArray("files");
                        JsonObject primaryFile = files.get(0).getAsJsonObject();
                        
                        for(JsonElement f : files) {
                            if (f.getAsJsonObject().has("primary") && f.getAsJsonObject().get("primary").getAsBoolean()) {
                                primaryFile = f.getAsJsonObject();
                                break;
                            }
                        }
                        
                        String url = primaryFile.get("url").getAsString();
                        String filename = primaryFile.get("filename").getAsString();
                        long size = primaryFile.get("size").getAsLong();
                        
                        sendStatus(sender, plugin.getMessage("status.downloading", Placeholder.unparsed("arg", filename)));
                        packageManager.downloadFile(url, packageManager.getPluginsDir(), filename, size, createProgressCallback(sender, filename));
                        sendStatus(sender, plugin.getMessage("status.downloaded", Placeholder.unparsed("arg", filename)));
                        
                    } catch (Exception e) {
                        sender.sendMessage(plugin.getMessage("errors.install-failed", Placeholder.unparsed("arg", slug), Placeholder.unparsed("error", e.getMessage())));
                    }
                }))
                .collect(Collectors.toList());

        CompletableFuture.allOf(installFutures.toArray(new CompletableFuture[0])).join();
        sender.sendMessage(plugin.getMessage("status.install-complete"));
    }

    private String getPkgName(JsonObject pkg) {
        if (pkg.has("slug")) return pkg.get("slug").getAsString();
        if (pkg.has("hangar_slug")) return pkg.get("hangar_slug").getAsString();
        return "unknown";
    }


    private Set<String> getInstalledProjectIds() {
        Map<String, String> installed = packageManager.getInstalledPlugins();
        Map<String, JsonObject> resolved = resolveVersions(new ArrayList<>(installed.values()));
        Set<String> ids = new HashSet<>();
        for (JsonObject v : resolved.values()) {
            ids.add(v.get("project_id").getAsString());
        }
        return ids;
    }

    private void resolveDependenciesRecursive(CommandSender sender, JsonObject project, Map<String, JsonObject> toInstall, Set<String> installedIds) {
        String id = project.get("id").getAsString();
        String slug = project.get("slug").getAsString();
        
        if (installedIds.contains(id) || toInstall.containsKey(id)) return;
        
        toInstall.put(id, project);
        
        try {
            JsonArray versions = ModrinthAPI.getVersions(id, List.of("spigot", "paper", "purpur", "bukkit"));
            if (versions.size() > 0) {
                JsonObject latest = versions.get(0).getAsJsonObject();
                JsonArray deps = latest.getAsJsonArray("dependencies");
                if (deps != null) {
                    for (JsonElement d : deps) {
                        JsonObject depObj = d.getAsJsonObject();
                        if (depObj.has("dependency_type") && depObj.get("dependency_type").getAsString().equals("required")) {
                            if (depObj.has("project_id") && !depObj.get("project_id").isJsonNull()) {
                                String depId = depObj.get("project_id").getAsString();
                                if (!installedIds.contains(depId) && !toInstall.containsKey(depId)) {
                                    sendStatus(sender, plugin.getMessage("status.checking", Placeholder.unparsed("arg", "dependency: " + depId)));
                                    JsonObject depProj = ModrinthAPI.getProject(depId);
                                    if (depProj != null) {
                                        resolveDependenciesRecursive(sender, depProj, toInstall, installedIds);
                                    }
                                }
                            }
                        }
                    }
                }
            }
        } catch (Exception e) {
            sender.sendMessage(plugin.getMessage("errors.failed-resolve-deps", Placeholder.unparsed("arg", slug), Placeholder.unparsed("error", e.getMessage())));
        }
    }
}

File: src/main/kotlin/io/github/Earth1283/aptMc/commands/sub/InstallCommand.kt
package io.github.Earth1283.aptMc.commands.sub

import com.google.gson.JsonObject
import io.github.Earth1283.aptMc.AptMc
import io.github.Earth1283.aptMc.api.GitHubAPI
import io.github.Earth1283.aptMc.api.HangarAPI
import io.github.Earth1283.aptMc.api.ModrinthAPI
import io.github.Earth1283.aptMc.commands.SubCommand
import io.github.Earth1283.aptMc.listeners.ConfirmationListener
import io.github.Earth1283.aptMc.managers.PackageManager
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder
import org.bukkit.command.CommandSender
import org.bukkit.configuration.file.YamlConfiguration
import java.io.File
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.util.concurrent.CompletableFuture

class InstallCommand(
    plugin: AptMc,
    packageManager: PackageManager,
    private val confirmationListener: ConfirmationListener?
) : SubCommand(plugin, packageManager) {

    override fun execute(sender: CommandSender, args: List<String>, dryRun: Boolean) {
        if (args.isEmpty()) {
            sender.sendMessage(plugin.getMessage("errors.no-packages-specified"))
            return
        }

        if (args.size == 1 && args[0].matches(Regex("^(http|https|ftp)://.*\\.yml$"))) {
            val url = args[0]
            if (dryRun) {
                sender.sendMessage(plugin.getMessage("status.dry-run", Placeholder.unparsed("arg", "Import manifest from $url")))
                return
            }
            try {
                sendStatus(sender, plugin.getMessage("status.downloading", Placeholder.unparsed("arg", "manifest")))
                val tempDir = File(plugin.dataFolder, "temp")
                if (!tempDir.exists()) tempDir.mkdirs()
                val tempFile = File(tempDir, "remote_import_" + System.currentTimeMillis() + ".yml")

                val client = HttpClient.newHttpClient()
                val request = HttpRequest.newBuilder()
                    .uri(URI.create(url))
                    .header("User-Agent", "apt-mc/1.2 (parody-cli)")
                    .GET()
                    .build()

                val response = client.send(request, HttpResponse.BodyHandlers.ofFile(tempFile.toPath()))

                if (response.statusCode() != 200) {
                    sender.sendMessage(plugin.getMessage("errors.fetch-manifest-failed", Placeholder.unparsed("error", "HTTP " + response.statusCode())))
                    return
                }

                if (!tempFile.exists() || tempFile.length() == 0L) {
                    sender.sendMessage(plugin.getMessage("errors.fetch-manifest-failed", Placeholder.unparsed("error", "File empty or missing")))
                    return
                }

                val yaml = YamlConfiguration.loadConfiguration(tempFile)
                val title = yaml.getString("project-details.title", "Unknown")!!
                val author = yaml.getString("project-details.author", "Unknown")!!

                sender.sendMessage(plugin.getMessage("status.import-header", Placeholder.unparsed("title", title), Placeholder.unparsed("author", author)))
                sender.sendMessage(plugin.getMessage("status.confirm-prompt"))

                if (confirmationListener != null) {
                    confirmationListener.addPending(sender, tempFile)
                } else {
                    sender.sendMessage(plugin.getMessage("errors.listener-missing"))
                }
            } catch (e: Exception) {
                sender.sendMessage(plugin.getMessage("errors.fetch-manifest-failed", Placeholder.unparsed("error", e.message ?: "")))
            }
            return
        }

        packageManager.ensureDir()
        sendStatus(sender, plugin.getMessage("status.update-reading"))
        sendStatus(sender, plugin.getMessage("status.update-building"))

        val installedIds = getInstalledProjectIds()
        val toInstallMap = LinkedHashMap<String, JsonObject>()
        var priority = plugin.getConfig().getStringList("source-priority")
        if (priority.isEmpty()) priority = listOf("modrinth", "hangar", "github")

        for (pkgSlug in args) {
            var found = false
            try {
                if (pkgSlug.contains("/") && !pkgSlug.startsWith("http")) {
                    sendStatus(sender, plugin.getMessage("status.checking", Placeholder.unparsed("arg", "$pkgSlug on GitHub")))
                    val release = GitHubAPI.getLatestRelease(pkgSlug)
                    if (release != null) {
                        val asset = GitHubAPI.getPrimaryJarAsset(release)
                        if (asset != null) {
                            val githubInfo = JsonObject()
                            githubInfo.addProperty("source", "github")
                            githubInfo.addProperty("slug", pkgSlug)
                            githubInfo.addProperty("download_url", asset.get("browser_download_url").asString)
                            githubInfo.addProperty("filename", asset.get("name").asString)
                            githubInfo.addProperty("size", asset.get("size").asLong)
                            toInstallMap["github:$pkgSlug"] = githubInfo
                            found = true
                        }
                    }
                }

                if (!found) {
                    for (source in priority) {
                        if (source.equals("modrinth", ignoreCase = true)) {
                            sendStatus(sender, plugin.getMessage("status.checking", Placeholder.unparsed("arg", "$pkgSlug on Modrinth")))
                            val project = ModrinthAPI.getProject(pkgSlug)
                            if (project != null) {
                                resolveDependenciesRecursive(sender, project, toInstallMap, installedIds)
                                found = true
                                break
                            }
                        } else if (source.equals("hangar", ignoreCase = true) && plugin.getConfig().getBoolean("enable-hangar", true)) {
                            sendStatus(sender, plugin.getMessage("status.checking", Placeholder.unparsed("arg", "$pkgSlug on Hangar")))
                            val hangarProject = HangarAPI.getProject(pkgSlug)
                            if (hangarProject != null) {
                                val versionInfo = HangarAPI.getLatestVersion(pkgSlug, "PAPER")
                                if (versionInfo != null) {
                                    versionInfo.addProperty("source", "hangar")
                                    versionInfo.addProperty("hangar_slug", hangarProject.get("name").asString)
                                    toInstallMap["hangar:" + hangarProject.get("name").asString] = versionInfo
                                    found = true
                                    break
                                }
                            }
                        }
                    }
                }

                if (!found) {
                    sender.sendMessage(plugin.getMessage("errors.package-not-found", Placeholder.unparsed("arg", pkgSlug)))
                }
            } catch (e: Exception) {
                sender.sendMessage(plugin.getMessage("errors.error-checking-pkg", Placeholder.unparsed("arg", pkgSlug), Placeholder.unparsed("error", e.message ?: "")))
            }
        }

        val toInstall = ArrayList(toInstallMap.values)
        if (toInstall.isEmpty()) {
            sender.sendMessage(plugin.getMessage("status.up-to-date"))
            return
        }

        if (dryRun) {
            sender.sendMessage(plugin.getMessage("status.dry-run-header"))
            for (pkg in toInstall) {
                val name = getPkgName(pkg)
                sender.sendMessage(plugin.getMessage("status.install-item", Placeholder.unparsed("arg", name)))
            }
            sender.sendMessage(plugin.getMessage("status.dry-run-complete"))
            return
        }

        if (plugin.getConfig().getBoolean("apt-song-references")) {
            sender.sendMessage(plugin.getMessage("status.install-song-ref"))
        }

        sender.sendMessage(plugin.getMessage("status.install-header"))
        for (pkg in toInstall) {
            val name = getPkgName(pkg)
            sender.sendMessage(plugin.getMessage("status.install-item", Placeholder.unparsed("arg", name)))
        }
        sender.sendMessage(plugin.getMessage("status.install-summary", Placeholder.unparsed("arg", toInstall.size.toString())))

        val installFutures = toInstall.map { pkg ->
            CompletableFuture.runAsync {
                val source = if (pkg.has("source")) pkg.get("source").asString else "modrinth"

                if (source == "github") {
                    try {
                        val url = pkg.get("download_url").asString
                        val filename = pkg.get("filename").asString
                        val size = pkg.get("size").asLong

                        sendStatus(sender, plugin.getMessage("status.downloading", Placeholder.unparsed("arg", filename)))
                        packageManager.downloadFile(url, packageManager.getPluginsDir(), filename, size, createProgressCallback(sender, filename))
                        sendStatus(sender, plugin.getMessage("status.downloaded", Placeholder.unparsed("arg", filename)))
                    } catch (e: Exception) {
                        sender.sendMessage(plugin.getMessage("errors.install-failed", Placeholder.unparsed("arg", pkg.get("slug").asString), Placeholder.unparsed("error", e.message ?: "")))
                    }
                    return@runAsync
                }

                if (source == "hangar") {
                    try {
                        val url = HangarAPI.getDownloadUrl(pkg, "PAPER")
                        val filename = HangarAPI.getFileName(pkg, "PAPER")

                        if (url == null || filename == null) {
                            sender.sendMessage(plugin.getMessage("errors.hangar-resolve-failed", Placeholder.unparsed("arg", pkg.get("hangar_slug").asString)))
                            return@runAsync
                        }

                        sendStatus(sender, plugin.getMessage("status.downloading", Placeholder.unparsed("arg", filename)))
                        packageManager.downloadFile(url, packageManager.getPluginsDir(), filename, -1, createProgressCallback(sender, filename))
                        sendStatus(sender, plugin.getMessage("status.downloaded", Placeholder.unparsed("arg", filename)))
                    } catch (e: Exception) {
                        sender.sendMessage(plugin.getMessage("errors.install-failed", Placeholder.unparsed("arg", "Hangar Package"), Placeholder.unparsed("error", e.message ?: "")))
                    }
                    return@runAsync
                }

                val slug = pkg.get("slug").asString
                try {
                    val versions = ModrinthAPI.getVersions(pkg.get("id").asString, listOf("spigot", "paper", "purpur", "bukkit"))
                    if (versions.size() == 0) {
                        sender.sendMessage(plugin.getMessage("errors.install-failed", Placeholder.unparsed("arg", slug), Placeholder.unparsed("error", "No compatible versions")))
                        return@runAsync
                    }

                    val latest = versions[0].asJsonObject
                    val files = latest.getAsJsonArray("files")
                    var primaryFile = files[0].asJsonObject

                    for (f in files) {
                        if (f.asJsonObject.has("primary") && f.asJsonObject.get("primary").asBoolean) {
                            primaryFile = f.asJsonObject
                            break
                        }
                    }

                    val url = primaryFile.get("url").asString
                    val filename = primaryFile.get("filename").asString
                    val size = primaryFile.get("size").asLong

                    sendStatus(sender, plugin.getMessage("status.downloading", Placeholder.unparsed("arg", filename)))
                    packageManager.downloadFile(url, packageManager.getPluginsDir(), filename, size, createProgressCallback(sender, filename))
                    sendStatus(sender, plugin.getMessage("status.downloaded", Placeholder.unparsed("arg", filename)))
                } catch (e: Exception) {
                    sender.sendMessage(plugin.getMessage("errors.install-failed", Placeholder.unparsed("arg", slug), Placeholder.unparsed("error", e.message ?: "")))
                }
            }
        }

        CompletableFuture.allOf(*installFutures.toTypedArray()).join()
        sender.sendMessage(plugin.getMessage("status.install-complete"))
    }

    private fun getPkgName(pkg: JsonObject): String {
        if (pkg.has("slug")) return pkg.get("slug").asString
        if (pkg.has("hangar_slug")) return pkg.get("hangar_slug").asString
        return "unknown"
    }

    private fun getInstalledProjectIds(): Set<String> {
        val installed = packageManager.getInstalledPlugins()
        val resolved = resolveVersions(ArrayList(installed.values))
        val ids = HashSet<String>()
        for (v in resolved.values) {
            ids.add(v.get("project_id").asString)
        }
        return ids
    }

    private fun resolveDependenciesRecursive(
        sender: CommandSender,
        project: JsonObject,
        toInstall: MutableMap<String, JsonObject>,
        installedIds: Set<String>
    ) {
        val id = project.get("id").asString
        val slug = project.get("slug").asString

        if (installedIds.contains(id) || toInstall.containsKey(id)) return

        toInstall[id] = project

        try {
            val versions = ModrinthAPI.getVersions(id, listOf("spigot", "paper", "purpur", "bukkit"))
            if (versions.size() > 0) {
                val latest = versions[0].asJsonObject
                val deps = latest.getAsJsonArray("dependencies")
                if (deps != null) {
                    for (d in deps) {
                        val depObj = d.asJsonObject
                        if (depObj.has("dependency_type") && depObj.get("dependency_type").asString == "required") {
                            if (depObj.has("project_id") && !depObj.get("project_id").isJsonNull) {
                                val depId = depObj.get("project_id").asString
                                if (!installedIds.contains(depId) && !toInstall.containsKey(depId)) {
                                    sendStatus(sender, plugin.getMessage("status.checking", Placeholder.unparsed("arg", "dependency: $depId")))
                                    val depProj = ModrinthAPI.getProject(depId)
                                    if (depProj != null) {
                                        resolveDependenciesRecursive(sender, depProj, toInstall, installedIds)
                                    }
                                }
                            }
                        }
                    }
                }
            }
        } catch (e: Exception) {
            sender.sendMessage(plugin.getMessage("errors.failed-resolve-deps", Placeholder.unparsed("arg", slug), Placeholder.unparsed("error", e.message ?: "")))
        }
    }
}
